flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/data/DataStructureConvertersTest.java

```diff
@@ -0,0 +1,617 @@
+
+			TestSpec
+				.forDataType(TIMESTAMP_WITH_LOCAL_TIME_ZONE(0))
+				.convertedTo(Instant.class, Instant.ofEpochSecond(12_345))
+				.convertedTo(Integer.class, 12_345)
+				.convertedTo(Long.class, 12_345_000L)
+				.convertedTo(TimestampData.class, TimestampData.fromEpochMillis(12_345_000)),
+
+			TestSpec
+				.forDataType(TIMESTAMP_WITH_LOCAL_TIME_ZONE(3))
+				.convertedTo(Instant.class, Instant.ofEpochSecond(12_345, 1_000_000))
+				.convertedTo(Long.class, 12_345_001L)
+				.convertedTo(TimestampData.class, TimestampData.fromEpochMillis(12_345_001)),
+
+			TestSpec
+				.forDataType(TIMESTAMP_WITH_LOCAL_TIME_ZONE(9))
+				.convertedTo(Instant.class, Instant.ofEpochSecond(12_345, 1))
+				.convertedTo(TimestampData.class, TimestampData.fromEpochMillis(12_345_000, 1)),
+
+			TestSpec
+				.forDataType(INTERVAL(YEAR(2), MONTH()))
+				.convertedTo(Period.class, Period.of(2, 6, 0))
+				.convertedTo(Integer.class, 30),
+
+			TestSpec
+				.forDataType(INTERVAL(MONTH()))
+				.convertedTo(Period.class, Period.of(0, 30, 0))
+				.convertedTo(Integer.class, 30),
+
+			TestSpec
+				.forDataType(INTERVAL(DAY(), SECOND(3))) // TODO support precision of 9
+				.convertedTo(Duration.class, Duration.ofMillis(123))
+				.convertedTo(Long.class, 123L),
+
+			TestSpec
+				.forDataType(ARRAY(BOOLEAN()))
+				.convertedTo(boolean[].class, new boolean[]{true, false, true, true})
+				.convertedTo(ArrayData.class, new GenericArrayData(new boolean[]{true, false, true, true})),
+
+			// arrays of TINYINT, SMALLINT, INTEGER, BIGINT, FLOAT, DOUBLE are skipped for simplicity
+
+			TestSpec
+				.forDataType(ARRAY(DATE()))
+				.convertedTo(
+					LocalDate[].class,
+					new LocalDate[]{null, LocalDate.parse("2010-11-12"), null, LocalDate.parse("2010-11-12")}),
+
+			TestSpec
+				.forDataType(MAP(INT(), BOOLEAN()))
+				.convertedTo(Map.class, createIdentityMap())
+				.convertedTo(MapData.class, new GenericMapData(createIdentityMap())),
+
+			TestSpec
+				.forDataType(MAP(DATE(), BOOLEAN()))
+				.convertedTo(Map.class, createLocalDateMap()),
+
+			TestSpec
+				.forDataType(MULTISET(BOOLEAN()))
+				.convertedTo(Map.class, createIdentityMultiset())
+				.convertedTo(MapData.class, new GenericMapData(createIdentityMultiset())),
+
+			TestSpec
+				.forDataType(MULTISET(DATE()))
+				.convertedTo(Map.class, createLocalDateMultiset()),
+
+			TestSpec
+				.forDataType(
+					ROW(
+						FIELD("a", INT()),
+						FIELD("b",
+							ROW(
+								FIELD("b_1", DOUBLE()),
+								FIELD("b_2", BOOLEAN())))))
+				.convertedTo(Row.class, Row.of(12, Row.of(2.0, null)))
+				.convertedTo(RowData.class, GenericRowData.of(12, GenericRowData.of(2.0, null))),
+
+			TestSpec
+				.forDataType(
+					ROW(
+						FIELD("a", INT()),
+						FIELD("b",
+							ROW(
+								FIELD("b_1", DATE()),
+								FIELD("b_2", DATE())))))
+				.convertedTo(Row.class, Row.of(12, Row.of(LocalDate.ofEpochDay(1), null))),
+
+			TestSpec
+				.forClass(PojoWithMutableFields.class)
```

Comment:
Could we add tests for pojo types nested in complex types (map, array, row). Those have the tendency to be error prone, as they have to properly open the nested converters. And the pojo converter is the only converter that has non trivial open method if I am not mistaken.